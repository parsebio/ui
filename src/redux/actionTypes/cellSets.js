const CELL_SETS = 'cellSets';

/**
 * Updates the cells sets shown to the user with the hierarchy fetched
 * from the API.
 */
const CELL_SETS_LOADED = `${CELL_SETS}/loaded`;

/**
 * Signals to the UI that new cell set data is about to be loaded in.
 */
const CELL_SETS_LOADING = `${CELL_SETS}/loading`;

/**
 * Creates a new cell set.
 */
const CELL_SETS_CREATE = `${CELL_SETS}/create`;

/**
 * Updates a single cell set's properties.
 */
const CELL_SETS_UPDATE_PROPERTY = `${CELL_SETS}/updateProperty`;

/**
 * Cell sets were reordered
 */
const CELL_SETS_REORDER = `${CELL_SETS}/reorder`;

/**
 * Deletes a cell set.
 */
const CELL_SETS_DELETE = `${CELL_SETS}/delete`;

/**
 * Deletes a cell class.
 */
const CELL_CLASS_DELETE = `${CELL_SETS}/deleteClass`;

/**
 * Saves all cell sets currently loaded.
 */
const CELL_SETS_SAVE = `${CELL_SETS}/save`;

/**
 * Hide a cell set from visualizations.
 */
const CELL_SETS_HIDE = `${CELL_SETS}/hide`;

/**
 * Unhide [remove a previous hidden state] from a cell set from visualizations.
 */
const CELL_SETS_UNHIDE = `${CELL_SETS}/unhide`;

/**
 * Unhide all hidden cell sets.
 */
const CELL_SETS_UNHIDE_ALL = `${CELL_SETS}/unhideAll`;

// /**
//  * Cell sets clustering being recalculated.
//  */
const CELL_SETS_CLUSTERING_UPDATING = `${CELL_SETS}/clusteringUpdating`;

// /**
//  * Cell sets clustering being recalculated.
//  */
const CELL_SETS_CLUSTERING_UPDATED = `${CELL_SETS}/clusteringUpdated`;

/**
 * Creates an error condition in the cell set tool.
 */
const CELL_SETS_ERROR = `${CELL_SETS}/error`;

export {
  CELL_SETS_LOADING, CELL_SETS_LOADED,
  CELL_SETS_CREATE,
  CELL_SETS_UPDATE_PROPERTY, CELL_SETS_REORDER,
  CELL_SETS_DELETE,
  CELL_CLASS_DELETE,
  CELL_SETS_SAVE,
  CELL_SETS_HIDE, CELL_SETS_UNHIDE, CELL_SETS_UNHIDE_ALL,
  CELL_SETS_CLUSTERING_UPDATING, CELL_SETS_CLUSTERING_UPDATED,
  CELL_SETS_ERROR,
};
