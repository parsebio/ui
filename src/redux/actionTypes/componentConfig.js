/**
 * Sets plot configs and data.
 */
const COMPONENT_CONFIG = 'componentConfig';
const LOAD_CONFIG = `${COMPONENT_CONFIG}/load`;
const UPDATE_CONFIG = `${COMPONENT_CONFIG}/update`;
const RESET_CONFIG = `${COMPONENT_CONFIG}/reset`;
const SAVE_CONFIG = `${COMPONENT_CONFIG}/save`;
const CONFIGS_REPLACED = `${COMPONENT_CONFIG}/replacements`;
const PLOT_DATA_LOADED = `${COMPONENT_CONFIG}/plotDataLoaded`;
const PLOT_DATA_LOADING = `${COMPONENT_CONFIG}/plotDataLoading`;
const PLOT_DATA_ERROR = `${COMPONENT_CONFIG}/plotDataError`;
const PLOT_DATA_CLEARED = `${COMPONENT_CONFIG}/plotDataCleared`;

const TRAJECTORY_NODES_SELECTION_UPDATED = `${COMPONENT_CONFIG}/trajectoryNodesUpdated`;

export {
  LOAD_CONFIG,
  UPDATE_CONFIG, CONFIGS_REPLACED,
  RESET_CONFIG, SAVE_CONFIG,
  PLOT_DATA_LOADED, PLOT_DATA_LOADING, PLOT_DATA_ERROR,
  TRAJECTORY_NODES_SELECTION_UPDATED, PLOT_DATA_CLEARED,
};
