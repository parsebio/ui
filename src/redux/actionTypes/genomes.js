const GENOMES = 'genomes';
const GENOME_FILES = 'genomeFiles';

const GENOME_FILE_UPDATE = `${GENOME_FILES}/update`;

const GENOMES_LOADING = `${GENOMES}/loading`;
const GENOMES_LOADED = `${GENOMES}/loaded`;
const GENOMES_UPDATED = `${GENOMES}/updated`;
const GENOMES_ERROR = `${GENOMES}/error`;

export {
  GENOME_FILE_UPDATE,
  GENOMES_LOADING,
  GENOMES_LOADED,
  GENOMES_ERROR,
  GENOMES_UPDATED,
};
