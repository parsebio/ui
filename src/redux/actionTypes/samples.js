const SAMPLES = 'samples';

/**
 * Samples created.
 */
const SAMPLES_CREATED = `${SAMPLES}/created`;

/**
 * Update sample.
 */
const SAMPLES_UPDATE = `${SAMPLES}/update`;

/**
 * Bulk Update sample kits.
 */
const SAMPLES_BULK_KIT_UPDATE = `${SAMPLES}/bulkKitUpdate`;

/**
 * Delete sample.
 */
const SAMPLES_DELETE = `${SAMPLES}/delete`;

/**
 * Update files in sample.
 */
const SAMPLES_FILE_UPDATE = `${SAMPLES}/fileUpdate`;

/**
 * Bulk update sample options.
 */
const SAMPLES_OPTIONS_UPDATE = `${SAMPLES}/optionsUpdate`;

/**
 * Load sample.
 */
const SAMPLES_LOADED = `${SAMPLES}/loaded`;

/**
 * Saving samples.
 */
const SAMPLES_SAVING = `${SAMPLES}/saving`;

/**
 * Samples saved successfully.
 */
const SAMPLES_SAVED = `${SAMPLES}/saved`;

/**
 * Error loading/saving sample.
 */
const SAMPLES_ERROR = `${SAMPLES}/error`;

/**
 * Delete metada from sample.
 */
const SAMPLES_METADATA_DELETE = `${SAMPLES}/metadataDelete`;

/**
 * Value in a metadata track was updated for a sample.
 */
const SAMPLES_VALUE_IN_METADATA_TRACK_UPDATED = `${SAMPLES}/valueInMetadataTrackUpdated`;

const SAMPLES_LOADING = `${SAMPLES}/loading`;

const SAMPLES_VALIDATING_UPDATED = `${SAMPLES}/validatingUpdated`;

export {
  SAMPLES_CREATED,
  SAMPLES_UPDATE,
  SAMPLES_BULK_KIT_UPDATE,
  SAMPLES_DELETE,
  SAMPLES_FILE_UPDATE,
  SAMPLES_OPTIONS_UPDATE,
  SAMPLES_LOADED,
  SAMPLES_SAVING,
  SAMPLES_ERROR,
  SAMPLES_SAVED,
  SAMPLES_METADATA_DELETE,
  SAMPLES_VALUE_IN_METADATA_TRACK_UPDATED,
  SAMPLES_LOADING,
  SAMPLES_VALIDATING_UPDATED,
};
