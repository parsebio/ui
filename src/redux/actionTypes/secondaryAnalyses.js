const SECONDARY_ANALYSES = 'secondaryAnalyses';

const SECONDARY_ANALYSES_CREATED = `${SECONDARY_ANALYSES}/created`;
const SECONDARY_ANALYSES_LOADING = `${SECONDARY_ANALYSES}/loading`;
const SECONDARY_ANALYSES_LOADED = `${SECONDARY_ANALYSES}/loaded`;
const SECONDARY_ANALYSES_ERROR = `${SECONDARY_ANALYSES}/error`;
const SECONDARY_ANALYSES_SAVING = `${SECONDARY_ANALYSES}/saving`;
const SECONDARY_ANALYSES_SET_ACTIVE = `${SECONDARY_ANALYSES}/setActive`;
const SECONDARY_ANALYSES_DELETED = `${SECONDARY_ANALYSES}/deleted`;
const SECONDARY_ANALYSES_UPDATED = `${SECONDARY_ANALYSES}/updated`;

const SECONDARY_ANALYSIS_FILES = 'secondaryAnalysisFiles';
const SECONDARY_ANALYSIS_FILES_LOADING = `${SECONDARY_ANALYSIS_FILES}/loading`;
const SECONDARY_ANALYSIS_FILES_LOADED = `${SECONDARY_ANALYSIS_FILES}/loaded`;
const SECONDARY_ANALYSIS_FILES_ERROR = `${SECONDARY_ANALYSIS_FILES}/error`;
const SECONDARY_ANALYSIS_FILES_UPDATE = `${SECONDARY_ANALYSIS_FILES}/update`;
const SECONDARY_ANALYSIS_FILES_CREATED = `${SECONDARY_ANALYSIS_FILES}/created`;
const SECONDARY_ANALYSIS_FILES_DELETE = `${SECONDARY_ANALYSIS_FILES}/delete`;

const SECONDARY_ANALYSIS_STATUS_LOADING = `${SECONDARY_ANALYSES}/executionLoading`;
const SECONDARY_ANALYSIS_STATUS_LOADED = `${SECONDARY_ANALYSES}/executionLoaded`;

export {
  SECONDARY_ANALYSES_CREATED,
  SECONDARY_ANALYSES_LOADING,
  SECONDARY_ANALYSES_LOADED,
  SECONDARY_ANALYSES_ERROR,
  SECONDARY_ANALYSES_SAVING,
  SECONDARY_ANALYSES_SET_ACTIVE,
  SECONDARY_ANALYSES_DELETED,
  SECONDARY_ANALYSES_UPDATED,
  SECONDARY_ANALYSIS_FILES_LOADING,
  SECONDARY_ANALYSIS_FILES_LOADED,
  SECONDARY_ANALYSIS_FILES_ERROR,
  SECONDARY_ANALYSIS_FILES_UPDATE,
  SECONDARY_ANALYSIS_FILES_CREATED,
  SECONDARY_ANALYSIS_FILES_DELETE,
  SECONDARY_ANALYSIS_STATUS_LOADING,
  SECONDARY_ANALYSIS_STATUS_LOADED,
};
