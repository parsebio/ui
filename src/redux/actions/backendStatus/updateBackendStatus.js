import {
  BACKEND_STATUS_UPDATED,
} from '../../actionTypes/backendStatus';

const updateBackendStatus = (experimentId, status) => async (dispatch) => {
  // Log the payload before dispatching the action
  console.log('Dispatching BACKEND_STATUS_UPDATED with payload:', { experimentId, status });

  dispatch({
    type: BACKEND_STATUS_UPDATED,
    payload: { experimentId, status },
  });
};

export default updateBackendStatus;
