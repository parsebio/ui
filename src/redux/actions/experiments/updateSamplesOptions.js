import endUserMessages from 'utils/endUserMessages';
import _ from 'lodash';

import {
  SAMPLES_OPTIONS_UPDATE, SAMPLES_SAVING, SAMPLES_ERROR,
} from 'redux/actionTypes/samples';

import handleError from 'utils/http/handleError';
import fetchAPI from 'utils/http/fetchAPI';
import { loadBackendStatus } from '../backendStatus';

const updateSamplesOptions = (experimentId, diff) => async (dispatch, getState) => {
  const url = `/v2/experiments/${experimentId}/samples/options`;

  const { sampleIds } = getState().experiments[experimentId];

  // The code assumes that the option of all samples are the same
  // This is the case until we support options at the sample level
  const oldOptions = getState().samples[sampleIds[0]].options;
  const newOptions = _.merge({}, oldOptions, diff);

  dispatch({
    type: SAMPLES_SAVING,
    payload: {
      message: endUserMessages.SAVING_SAMPLE,
    },
  });

  try {
    await fetchAPI(
      url,
      {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(newOptions),
      },
    );

    dispatch({
      type: SAMPLES_OPTIONS_UPDATE,
      payload: {
        sampleUuids: sampleIds,
        diff,
      },
    });

    await dispatch(loadBackendStatus(experimentId));
  } catch (e) {
    const errorMessage = handleError(e, endUserMessages.ERROR_SAVING);

    dispatch({
      type: SAMPLES_ERROR,
      payload: {
        error: errorMessage,
      },
    });
  }
};

export default updateSamplesOptions;
