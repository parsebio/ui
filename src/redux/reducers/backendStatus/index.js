import initialState from './initialState';

import {
  BACKEND_STATUS_LOADING,
  BACKEND_STATUS_LOADED,
  BACKEND_STATUS_UPDATED,
  BACKEND_STATUS_ERROR,
} from '../../actionTypes/backendStatus';
import { EXPERIMENTS_DELETED } from '../../actionTypes/experiments';

import experimentsDeleted from './experimentsDeleted';
import backendStatusLoading from './backendStatusLoading';
import backendStatusLoaded from './backendStatusLoaded';
import backendStatusUpdated from './backendStatusUpdated';
import backendStatusError from './backendStatusError';

const backendStatusReducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case BACKEND_STATUS_LOADING: {
      return backendStatusLoading(state, action);
    }
    case BACKEND_STATUS_LOADED: {
      return backendStatusLoaded(state, action);
    }
    case BACKEND_STATUS_UPDATED: {
      return backendStatusUpdated(state, action);
    }
    case BACKEND_STATUS_ERROR: {
      return backendStatusError(state, action);
    }
    case EXPERIMENTS_DELETED: {
      return experimentsDeleted(state, action);
    }
    default: {
      return state;
    }
  }
};

export default backendStatusReducer;
