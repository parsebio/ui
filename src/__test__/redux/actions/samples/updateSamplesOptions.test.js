import _ from 'lodash';
import configureStore from 'redux-mock-store';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import thunk from 'redux-thunk';

import updateSamplesOptions from 'redux/actions/experiments/updateSamplesOptions';
import initialSamplesState, { sampleTemplate } from 'redux/reducers/samples/initialState';
import initialExperimentState, { experimentTemplate } from 'redux/reducers/experiments/initialState';

import {
  SAMPLES_OPTIONS_UPDATE,
  SAMPLES_ERROR, SAMPLES_SAVING,
} from 'redux/actionTypes/samples';
import { BACKEND_STATUS_LOADED, BACKEND_STATUS_LOADING } from 'redux/actionTypes/backendStatus';

const mockStore = configureStore([thunk]);

const mockExperimentId = 'experiment-1';
const mockSampleIds = ['sample-1', 'sample-2'];

const mockSamples = mockSampleIds.reduce((acc, sampleId, idx) => {
  acc[sampleId] = {
    ...sampleTemplate,
    name: `test sample ${idx}`,
    uuid: `sampleId-${idx}`,
  };
  return acc;
}, {});

const mockState = {
  experiments: {
    ...initialExperimentState,
    ids: [mockExperimentId],
    [mockExperimentId]: {
      ...experimentTemplate,
      name: 'Mock experiment 1',
      sampleIds: mockSampleIds,
    },
    meta: {
      ...initialExperimentState.meta,
      activeExperimentId: mockExperimentId,
    },
  },
  samples: {
    ...initialSamplesState,
    ...mockSamples,
  },
};

const endpointUrl = `http://localhost:3000/v2/experiments/${mockExperimentId}/samples/options`;

describe('updateSamplesOptions action', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    enableFetchMocks();
    fetchMock.resetMocks();
    fetchMock.doMock();
  });

  it('Works correctly', async () => {
    fetchMock
      .mockResponseOnce(() => Promise.resolve(JSON.stringify({})))
      .mockResponseOnce(() => Promise.resolve(JSON.stringify({})));

    const sampleDiff = {
      includeAbSeq: false,
    };

    const store = mockStore(mockState);

    await store.dispatch(updateSamplesOptions(mockExperimentId, sampleDiff));

    const actions = store.getActions();
    expect(_.map(actions, 'type')).toEqual([SAMPLES_SAVING, SAMPLES_OPTIONS_UPDATE, BACKEND_STATUS_LOADING, BACKEND_STATUS_LOADED]);
    expect(_.map(actions, 'payload')).toMatchSnapshot();

    expect(fetchMock).toHaveBeenCalledWith(
      endpointUrl,
      {
        body: JSON.stringify(sampleDiff),
        headers: { 'Content-Type': 'application/json' },
        method: 'PUT',
      },
    );
  });

  it('Error handling works', async () => {
    fetchMock.mockRejectOnce(() => Promise.reject(new Error('Api error')));

    const sampleDiff = {
      name: 'updated name',
    };

    const store = mockStore(mockState);
    await store.dispatch(updateSamplesOptions(mockExperimentId, sampleDiff));

    const actions = store.getActions();
    expect(_.map(actions, 'type')).toEqual([SAMPLES_SAVING, SAMPLES_ERROR]);
    expect(_.map(actions, 'payload')).toMatchSnapshot();
  });
});
