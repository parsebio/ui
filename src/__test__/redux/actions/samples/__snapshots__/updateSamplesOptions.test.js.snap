// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateSamplesOptions action Error handling works 1`] = `
[
  {
    "message": "Saving sample...",
  },
  {
    "error": "We couldn't save your data.",
  },
]
`;

exports[`updateSamplesOptions action Works correctly 1`] = `
[
  {
    "message": "Saving sample...",
  },
  {
    "diff": {
      "includeAbSeq": false,
    },
    "sampleUuids": [
      "sample-1",
      "sample-2",
    ],
  },
  {
    "experimentId": "experiment-1",
  },
  {
    "experimentId": "experiment-1",
    "status": {},
  },
]
`;
